### PR TITLE
BUG: is_string_dtype(pd.Index([], dtype='O')) returns False

### LINKED ISSUE #54661: BUG: is_string_dtype(pd.Index([], dtype='O')) returns `False`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.
- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
from pandas.api.types import is_string_dtype

assert is_string_dtype(pd.Index([], dtype='O'))
```

### Issue Description

For object arrays it tests if all objects are strings. That test should be successful when 0 items are tested.

### Expected Behavior

The assertion succeeds.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : 6e194cf9eadedbf5506da5ba0939d700f46ba3ba
python              : 3.10.12.final.0
python-bits         : 64
OS                  : Linux
OS-release          : 6.4.10-zen2-1-zen
Version             : #1 ZEN SMP PREEMPT_DYNAMIC Sun, 13 Aug 2023 01:33:59 +0000
machine             : x86_64
processor           : 
byteorder           : little
LC_ALL              : None
LANG                : de_DE.UTF-8
LOCALE              : de_DE.UTF-8

pandas              : 2.1.0rc0
numpy               : 1.24.4
pytz                : 2023.3
dateutil            : 2.8.2
setuptools          : 68.1.0
pip                 : 23.2.1
Cython              : None
pytest              : 7.4.0
hypothesis          : None
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : 3.1.2
IPython             : None
pandas_datareader   : None
bs4                 : None
bottleneck          : None
brotli              : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : 2023.6.0
gcsfs               : None
matplotlib          : 3.7.2
numba               : 0.57.1
numexpr             : None
odfpy               : None
openpyxl            : 3.1.2
pandas_gbq          : None
pyarrow             : None
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.11.2
snappy              : None
sqlalchemy          : None
tables              : None
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : None
pyqt5               : None

</details>

### PR DESCRIPTION
**Pull request by natmokval**
- [x] closes #54661
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit)
- [x] Added an entry in `doc/source/whatsnew/v2.2.0.rst`

Corrected the definition of `is_all_strings`. Now for object arrays, which have no items `is_string_dtype` returns `True` instead of `False`.

### CODE REVIEW

File: pandas/tests/dtypes/test_dtypes.py [hunk 1]
```diff
@@ -1212,3 +1212,8 @@ def test_multi_column_dtype_assignment():
 
     df["b"] = 0
     tm.assert_frame_equal(df, expected)
+
+
+def test_empty_object_array_is_string_dtype():
+    # GH #54661
```
Comment (jbrockmendel) [inline pandas/tests/dtypes/test_dtypes.py:1218]: nitpick can you remove the space after GH.  i like the pattern GH#12345 bc it is something i grep for

File: pandas/tests/dtypes/test_dtypes.py [hunk 2]
```diff
@@ -1212,3 +1212,8 @@ def test_multi_column_dtype_assignment():
 
     df["b"] = 0
     tm.assert_frame_equal(df, expected)
+
+
+def test_empty_object_array_is_string_dtype():
```
Comment (jbrockmendel) [inline pandas/tests/dtypes/test_dtypes.py:1217]: hmm looks like we have tests for is_string_dtype in a couple of places.  one of these days might be worth checking if there is a reason for that and if not, refactoring

File: doc/source/whatsnew/v2.2.0.rst [hunk 3]
```diff
@@ -169,6 +169,7 @@ Bug fixes
 ~~~~~~~~~
 - Bug in :class:`AbstractHolidayCalendar` where timezone data was not propagated when computing holiday observances (:issue:`54580`)
 - Bug in :class:`pandas.core.window.Rolling` where duplicate datetimelike indexes are treated as consecutive rather than equal with ``closed='left'`` and ``closed='neither'`` (:issue:`20712`)
+- Bug in :func:`is_all_strings` while checking object array with no elements is of the string dtype (:issue:`54661`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.2.0.rst:172]: is the bug in is_all_strings or is_string_dtype?

File: pandas/tests/dtypes/test_dtypes.py (hunk 2 above)
Comment (natmokval) [reply pandas/tests/dtypes/test_dtypes.py]: Thank you for the comment. I think, I found a better place for my test: `pandas/tests/dtypes/test_common.py`. I combined my test with another one and added a parameterization.

File: doc/source/whatsnew/v2.2.0.rst (hunk 3 above)
Comment (natmokval) [reply doc/source/whatsnew/v2.2.0.rst]: I think, the bug is in `is_all_strings`, but we don’ t have this function in api and users see the bug while calling `is_string_dtype`. That is why I replaced Bug in :func:`is_all_strings` with: Bug in :func:`is_string_dtype`.

Looks like we don’t have a test for is_all_strings, do we need one?

File: pandas/tests/dtypes/test_dtypes.py (hunk 1 above)
Comment (natmokval) [reply pandas/tests/dtypes/test_dtypes.py]: Done. I agree, it's better to have all GH refers look alike

File: doc/source/whatsnew/v2.2.0.rst (hunk 3 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v2.2.0.rst]: i meant the user-facing bug.  in the docs we generally only reference public functions/methods

Comment (natmokval) [reply doc/source/whatsnew/v2.2.0.rst]: okay, I removed from `doc/source/whatsnew/v2.2.0.rst` my line.

Comment (natmokval) [reply doc/source/whatsnew/v2.2.0.rst]: @jbrockmendel, I updated the PR. Could you please take a look at it.

Comment (natmokval) [reply doc/source/whatsnew/v2.2.0.rst]: Yeah, `is_string_dtype` is a public function, I added the note to `v2.1.1.rst`. I think my change is a minor and we can put it in the version 2.1.1

File: doc/source/whatsnew/v2.1.1.rst [hunk 4]
```diff
@@ -34,6 +34,7 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
 - Fixed bug for :class:`ArrowDtype` raising ``NotImplementedError`` for fixed-size list (:issue:`55000`)
+- Fixed bug in :func:`is_string_dtype` while checking object array with no elements is of the string dtype (:issue:`54661`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.1.1.rst:37]: I think you'll need this to be
```
:func:`pandas.api.types.is_string_dtype`
```
and to move this to 2.2

**Comment by natmokval** [comment]
@jbrockmendel, I corrected my PR, could you please take a look at the update?

**Review by jbrockmendel — APPROVED** [review]

**Comment by jbrockmendel** [comment]
LGTM.  pls merge main and ping on green

**Comment by natmokval** [comment]
> LGTM. pls merge main and ping on green

thanks, I merged main, ci - green

**Pull request merged**